manticore/ethereum/detectors.py

```diff
@@ -577,3 +577,112 @@ def did_evm_read_storage_callback(self, state, address, offset, value):
     def did_evm_write_storage_callback(self, state, address, offset, value):
         # concrete or symbolic write
         state.context.setdefault('{:s}.initialized_storage'.format(self.name), set()).add((address, offset))
+
+
+class DetectRaceCondition(Detector):
+    """Detects possible transaction race conditions (transaction order dependencies)
+
+    The RaceCondition detector might not work properly for contracts that have only a fallback function.
+    See the detector's implementation and it's `_in_user_func` method for more information.
+    """
+
+    TAINT = 'written_storage_slots.'
+
+    def __init__(self, *a, **kw):
+        # Normally `add_finding_here` makes it unique reporting but
+        # we might try to report the same thing multiple times e.g. in consecutive instructions
+        # so we need to make our own 'unique findings' set too.
+        self.__findings = set()
+        super().__init__(*a, **kw)
+
+    @staticmethod
+    def _in_user_func(state):
+        """
+        :param state: current state
+        :return: whether the current execution is in a user-defined function or not.
+
+        NOTE / TODO / FIXME: As this may produce false postives, this is not in the base `Detector` class.
+        It should be fixed at some point and moved there. See below.
+
+        The first 4 bytes of tx data is keccak256 hash of the function signature that is called by given tx.
+
+        All transactions start within Solidity dispatcher function: it takes passed hash and dispatches
+        the execution to given function based on it.
+
+        So: if we are in the dispatcher, *and contract have some functions* one of the first four tx data bytes
+        will effectively have more than one solutions.
+
+        BUT if contract have only a fallback function, the equation below may return more solutions when we are
+        in a dispatcher function.  <--- because of that, we warn that the detector is not that stable
+        for contracts with only a fallback function.
+        """
+
+        # TODO / FIXME: Benchmark/check if doing it byte by byte and also returning the bytes is faster
+        return len(state.solve_n(state.platform.current_transaction.data[:4], 2)) == 1
```

Comment:
@feliam we were speaking about caching this result; what would be the proper cache key? Current pc? Is that okay to take a hash out of it then as it may be e.g. a `BitVecAdd`?